src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/PkcsPalOpenSsl.Encrypt.cs

```diff
@@ -70,14 +71,94 @@ internal sealed partial class PkcsPalOpenSsl : PkcsPal
                     if (unprotectedAttributes.Count != 0)
                     {
                         // TODO(3334): Add support for unprotected attributes here.
-                        throw new NotImplementedException();
+                        encryptedMessage = AddAttributesToEncoding(encryptedMessage, unprotectedAttributes);
                     }
 
                     return encryptedMessage;
                 }
             }
         }
 
+        private byte[] AddAttributesToEncoding(byte[] encryptedMessage, CryptographicAttributeObjectCollection unprotectedAttributes)
+        {
+            DerSequenceReader originalMessage = new DerSequenceReader(encryptedMessage);
+
+            // EnvelopedCms has the following underlying structure:
+            //
+            // ContentInfo ::= SEQUENCE {
+            //      contentType ContentType,
+            //      content[0] EXPLICIT ANY DEFINED BY contentType }
+            //
+            // ContentType ::= OBJECT IDENTIFIER
+            //
+            // According to RFC2630, the enveloped type we have support for has the following structure
+            // EnvelopedData ::= SEQUENCE {
+            //     version CMSVersion,
+            //     originatorInfo[0] IMPLICIT OriginatorInfo OPTIONAL,
+            //     recipientInfos RecipientInfos,
+            //     encryptedContentInfo EncryptedContentInfo,
+            //     unprotectedAttrs[1] IMPLICIT UnprotectedAttributes OPTIONAL }
+
+            byte[][] contentType = originalMessage.ReadAndSplitNextEncodedValue();
+            DerSequenceReader envelopedDataReader = originalMessage.ReadExplicitContextSequence();
+
+            byte[][] version = envelopedDataReader.ReadAndSplitNextEncodedValue();
+
+            const byte ConstructedContextSpecificTag = DerSequenceReader.ContextSpecificTagFlag | DerSequenceReader.ConstructedFlag;
+
+            byte[][] originatorInfo = (envelopedDataReader.PeekTag() == ConstructedContextSpecificTag) ?
+                originatorInfo = envelopedDataReader.ReadAndSplitNextEncodedValue() :
+                new byte[][] { Array.Empty<byte>(), Array.Empty<byte>(), Array.Empty<byte>() };
```

Comment:
If you need to keep extracting semantic values, consider saving this array as a static readonly to save on allocations